mfsmaster/chunks.c

```diff
@@ -2861,6 +2862,123 @@ void chunk_store_info(uint8_t *buff) {
 
 //jobs state: jobshpos
 
+static inline uint16_t chunk_get_undergoal_replicate_srccsid(chunk *c, uint16_t dstcsid, uint32_t now, uint32_t lclass, uint32_t rgvc, uint32_t rgtdc) {
+	slist *s;
+	uint32_t r = 0;
+	uint16_t srccsid = MAXCSCOUNT;
+
+	if (ReplicationsRespectTopology) {
+		uint32_t min_dist = 0xFFFFFFFF;
+		uint32_t dist;
+		uint32_t ip;
+		uint32_t cuip;
+		uint32_t mdrgvc = 0;
+		uint32_t mdrgtdc = 0;
+
+		if (matocsserv_get_csdata(cstab[dstcsid].ptr,&cuip,NULL,NULL,NULL)==0) {
+			for (s=c->slisthead ; s ; s=s->next) {
+				if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && (s->valid==VALID || s->valid==TDVALID)) {
+					if (matocsserv_get_csdata(cstab[s->csid].ptr,&ip,NULL,NULL,NULL)==0) {
+						dist=topology_distance(ip,cuip);
+						if (min_dist>=dist) {
+							if (min_dist>dist) {
+								min_dist=dist;
+								srccsid=s->csid;
+								mdrgvc = 0;
+								mdrgtdc = 0;
+							} else if (s->valid==VALID) {
+								srccsid=s->csid;
+							}
+							if (s->valid==VALID) {
+								mdrgvc++;
+							} else {
+								mdrgtdc++;
+							}
+						}
+					}
+				}
+			}
+			if (mdrgvc > 1) {
```

Comment:
Hi @acid-maker 

> you shouldnt set "srccsid" in the first pass - you should only find min_dist and element counts ("mdrgvc" and "mdrftdc"). "srccsid" should be set only in the second pass

If there're no more than 1 VALID copy or TDVALID copy, we don't need to randomize it, so no need to go through the second `for` loop - and the `r` will be 0, for loop condition is `s && r>0`. So I set `srccsid` in the first loop for this case.

Reply:
ok.
BTW. It is possible to randomize it using only one pass.